### ISSUE TITLE
Linux : missing vulkan backend support

### ISSUE THREAD

**Issue by mob41** [RPCS3/rpcs3#1752]
As from pull-request https://github.com/RPCS3/rpcs3/pull/1751#issuecomment-223826063

**Comment by rcaridade145** [comment]
The m_handle on Linux , better wxGTK returns GTKWidget as per - http://docs.wxwidgets.org/trunk/classwx_window.html#a185e6cd7065367b552748cb722651b27 .
This means that depending on the backend of WxWidgets we have to use the correct API to retrieve the X11 Window.
For instance on GTK3:

```
GtkWidget *widget =(GtkWidget*)m_frame->handle();
Window w = gdk_x11_window_get_xid((gtk_widget_get_window(widget));
```

This means depending on GTK and on its paths as a dependency. Is this ok? Still need to overcome this, however it is coming along i guess.

**Comment by mob41** [comment]
I am still wondering that wxWidget's wxWindow will return a proper Handle (```HWND``` in Windows). So I am still setting up my Linux machine. Phew! I have to reinstall it for wrong architecture.

The function ```GetHandle()``` seems not returning a valid one. Any ideas?

**Comment by mob41** [comment]
@rcaridade145 And why ```GtkWidget```? Unless you want to do a rewrite of the GUI for Linux, you think you don't need to use another library to do this. wxWidget has its own ```handle``` (HWND) to get from. But I still need to compile it on Linux, Windows to see is it compatible for all platforms.

**Comment by rcaridade145** [comment]
@mob41 GTK is the backend of wxWidgets on Linux. I've pushed a work in progress to my branch, not production ready, still needs some work. Getting

 `/home/rui/Programacao/rpcs3/rpcs3/Emu/Cell/Modules/cellPngDec.cpp: In function ‘be_t<unsigned int> pngDecGetChunkInformation(PStream, bool)’:
/home/rui/Programacao/rpcs3/rpcs3/Emu/Cell/Modules/cellPngDec.cpp:212:98: error: invalid conversion from ‘png_byte** {aka unsigned char**}’ to ‘png_charpp {aka char**}’ [-fpermissive]
 ream->png_ptr, stream->info_ptr, &name, &compression_type, &profile, &proflen))
                                                                              ^
/home/rui/Programacao/rpcs3/rpcs3/Emu/Cell/Modules/cellPngDec.cpp:212:98: error: cannot convert ‘u32* {aka unsigned int*}’ to ‘png_uint_32* {aka long unsigned int*}’ for argument ‘6’ to ‘png_uint_32 png_get_iCCP(png_structp, png_infop, png_charpp, int*, png_charpp, png_uint_32*)’
/home/rui/Programacao/rpcs3/rpcs3/Emu/Cell/Modules/cellPngDec.cpp: In function ‘s32 pngDecodeData(PPUThread&, PHandle, PStream, vm::ps3::ptr<unsigned char>, PDataControlParam, PDataOutInfo, PCbControlDisp, PDispParam)’:
/home/rui/Programacao/rpcs3/rpcs3/Emu/Cell/Modules/cellPngDec.cpp:611:40: error: ‘PNG_ERROR_ACTION_NONE’ was not declared in this scope
   png_set_rgb_to_gray(stream->png_ptr, PNG_ERROR_ACTION_NONE, PNG_RGB_TO_GRAY_D
                                        ^
/home/rui/Programacao/rpcs3/rpcs3/Emu/Cell/Modules/cellPngDec.cpp:611:63: error: ‘PNG_RGB_TO_GRAY_DEFAULT’ was not declared in this scope
   png_set_rgb_to_gray(stream->png_ptr, PNG_ERROR_ACTION_NONE, PNG_RGB_TO_GRAY_D
                                                               ^
rpcs3/CMakeFiles/rpcs3.dir/build.make:5926: recipe for target 'rpcs3/CMakeFiles/rpcs3.dir/Emu/Cell/Modules/cellPngDec.cpp.o' failed
make[2]: *** [rpcs3/CMakeFiles/rpcs3.dir/Emu/Cell/Modules/cellPngDec.cpp.o] Error 1
`
when trying to compile later on. Need to investigate.

**Comment by mob41** [comment]
However I have no freetime taking care about graphics. At least I still have to continue my development of "Joystick support" and "Mixed(kb+joy) support". But they have relationship will the window handle. Ar.. 1 more week to go through of my academic exam...

**Comment by Nezarn** [comment]
@rcaridade145 have you tried taking a look at #1687 ? it includes fixes for building in linux.

**Comment by rcaridade145** [comment]
@Nezarn It seems that it might help indeed due to changes on libpng.

**Comment by rcaridade145** [comment]
```
CMakeFiles/rpcs3.dir/Emu/RSX/VK/VKGSRender.cpp.o: na função `vk::context::createSwapChain(unsigned long, vk::physical_device&)':
VKGSRender.cpp:(.text._ZN2vk7context15createSwapChainEmRNS_15physical_deviceE[_ZN2vk7context15createSwapChainEmRNS_15physical_deviceE]+0x84): referência indefinida a `vkCreateXlibSurfaceKHR'
collect2: error: ld returned 1 exit status

```

I'm having troubles linking the code. I see that on the latest version at https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/blob/master/CMakeLists.txt xlib support is on. Can we update that submodule? @tambry  @raven02

**Comment by tambry** [comment]
I'm no longer part of the development team, please avoid pinging me in the future. We couldn't update that submodule due to our Travis buildbots being on a very old distro, but seems like Trusty is now used so it should be possible to update it. Do note that it needs quite a few new dependencies to build on Linux on a newer version.

**Comment by GhettoGirl** [comment]
> We couldn't update that submodule due to our Travis buildbots being on a very old distro

Just a suggestion: Use **[Arch Travis](https://github.com/mikkeloscar/arch-travis)** instead of Ubuntu. Arch Linux is always up to date. Don't bother with setting up external repositories and binaries. For one of my projects I could reduce the build time from 15 minutes to 2 minutes. Also the default repository of Arch Travis is very fast.

**Comment by rcaridade145** [comment]
Got a little time this weekend ad reworked the code to use XCB instead of XLIB.
It does compile and start  however everything is blank.
![blankscreen](https://cloud.githubusercontent.com/assets/11843298/16547555/26ab8d6c-416a-11e6-93cd-3e96504beecb.jpg)
Will have to look into it a bit more.

**Comment by GhettoGirl** [comment]
@rcaridade145 have you tried to compile with ```cmake -DUSE_SYSTEM_FFMPEG=ON``` at least for me the builtin patched ppsspp-ffmpeg breaks the rendering window.
I guess this will fix this.

**Comment by rcaridade145** [comment]
An update on this.. After Nvidia launched 370 series i'm starting to see more stuff appearing on Linux with Vulkan

![vulkan_linux](https://cloud.githubusercontent.com/assets/11843298/17834424/98b4af36-6737-11e6-9d9b-7fe8173c9cc3.png)

Some errors do appear on the console

`E {rsx::thread} RSX: ERROR: [MEM] Code 16 : vkCmdBeginRenderPass(): Cannot read invalid memory 7fbebc002650, please fill the memory before using.
E {rsx::thread} RSX: ERROR: [DS] Code 7 : Cannot submit cmd buffer using image (7fbebc0025a0) with layout VK_IMAGE_LAYOUT_UNDEFINED when first use is VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL.
E {rsx::thread} RSX: ERROR: [DS] Code 7 : Cannot submit cmd buffer using image (7fbebc0025a0) with layout VK_IMAGE_LAYOUT_UNDEFINED when first use is VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL.
E {rsx::thread} RSX: ERROR: [DS] Code 7 : Cannot submit cmd buffer using image (7fbebc007a90) with layout VK_IMAGE_LAYOUT_UNDEFINED when first use is VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL.
`
@raven02 Issues with NVIDIA on Linux or more general?

**Comment by kd-11** [comment]
You should consider submitting a PR with this work for initial commit. I can clean up any remaining issues afterwards. I'm now working on getting linux stuff to catch up to windows.

**Comment by exolyte** [comment]
I've tested the vulkan build for a bit and it seems like the keyboard input is the only major issue. I haven't looked into it too much, but this could be a gtk/wxWidgets issue so it might be better to wait till the qt port is complete.

**Comment by kd-11** [comment]
Its unlikely to be an issue with gtk/wx pretty sure its somehow rpcs3 doing something shady. Either way, waiting for QT means the QT port will not have the xcb stuff and we'll need to wait for the author to update the code again. I'd rather we just work on it now, so that its ready when the QT work is merged or shortly thereafter.

**Comment by rcaridade145** [comment]
I'm trying to get it to build. Last versions with atomic broke the build for me.

> pretty sure its somehow rpcs3 doing something shady

Or not doing something at all :)

**Comment by kd-11** [comment]
I thought the atomic issue was gcc 7 only? Also, opening a PR helps move things along and vulkan on linux cannot proceed until we do so. Just check that it builds on travis.

**Comment by rcaridade145** [comment]
> I thought the atomic issue was gcc 7 only? 

`std::atomic<int> m_queued_threads = 0;`

Add to replace it with

`std::atomic<int> m_queued_threads = {0};`

As found here http://stackoverflow.com/questions/27314485/use-of-deleted-function-error-with-stdatomic-int

**Comment by AniLeo** [comment]
Solved by #2842
